proxylib/proxylib.go

```diff
@@ -63,61 +58,14 @@ func OnNewConnection(instanceId uint64, proto string, connectionId uint64, ingre
 	if instance == nil {
 		return C.FILTER_INVALID_INSTANCE
 	}
-	// Find the parser for the proto
-	parserFactory := GetParserFactory(proto)
-	if parserFactory == nil {
-		return C.FILTER_UNKNOWN_PARSER
-	}
-	_, port, err := net.SplitHostPort(dstAddr)
-	if err != nil {
-		return C.FILTER_INVALID_ADDRESS
-	}
-	dstPort, err := strconv.ParseUint(port, 10, 32)
-	if err != nil || dstPort == 0 {
-		return C.FILTER_INVALID_ADDRESS
-	}
-	// Note: Strings passed as arguments from C must be copied to Go memory to keep them valid after this function
-	// returns
-	connection := &Connection{
-		Instance:   instance,
-		Id:         connectionId,
-		Ingress:    ingress,
-		SrcId:      srcId,
-		DstId:      dstId,
-		SrcAddr:    strcpy(srcAddr),
-		DstAddr:    strcpy(dstAddr),
-		Port:       uint32(dstPort),
-		PolicyName: strcpy(policyName),
-		ParserName: strcpy(proto),
-		OrigBuf:    origBuf,
-		ReplyBuf:   replyBuf,
-	}
-	connection.Parser = parserFactory.Create(connection)
-	if connection.Parser == nil {
-		// Parser rejected the new connection based on the connection metadata
-		return C.FILTER_POLICY_DROP
-	}
 
-	mutex.Lock()
-	connections[connectionId] = connection
-	mutex.Unlock()
-
-	return C.FILTER_OK
-}
-
-// Skip bytes in input, or exhaust the input.
-func advanceInput(input [][]byte, bytes int) [][]byte {
-	for bytes > 0 && len(input) > 0 {
-		rem := len(input[0]) // this much data left in the first slice
-		if bytes < rem {
-			input[0] = input[0][bytes:] // skip 'bytes' bytes
-			bytes = 0
-		} else { // go to the beginning of the next unit
-			bytes -= rem
-			input = input[1:] // may result in an empty slice
-		}
+	res, conn := NewConnection(instance, strcpy(proto), connectionId, ingress, srcId, dstId, strcpy(srcAddr), strcpy(dstAddr), strcpy(policyName), origBuf, replyBuf)
```

Comment:
Cgo allow strings and slices of trivial types to be passed from C to Go and Cgo defines the C-types for string and slice headers. I my experience Cgo does not (thankfully) copy the underlying data. Cgo documentation explicitly states that C can't hold references to Go memory (due to GC). It is not so strict in the other direction, but generally the C side needs to keep any memory passed to Go valid, if references are retained longer than for the duration of the function call.

In this specific case we are passing strings from C to Go. The problem is that I have no provisions on the calling side to keep any of the underlying memory around longer than the duration of the function call, so the data might be in the stack, etc.

According to Go documentation the underlying data is copied whenever a string is cast/sliced to a slice of bytes, and the other way around. My definition of `strcpy()` uses this and first casts  the string as a slice of bytes, and then creates a string using those bytes. In my experience this works. However, maybe it would be safer (long term) for `strcpy()` to use `copy()` explicitly for the first copy, and then cast it back to string?

Reply:
Oh, I meant using `func C.GoString(*C.char) string` which apparently makes a copy:
 "A few special functions convert between Go and C types by making copies of the data" (https://golang.org/cmd/cgo/)
but I actually don't fully trust that ;) What you have does the same thing.
I had read the strcpy docstring as making a new copy every time but, reading it again, that's clearly not the expectation (just a copy from the C). Anyway, not a big deal!